sources/lib/frontend/typechecker.ml

```diff
@@ -120,24 +125,38 @@ module Types = struct
             error (UnknownType s) loc
       end
 
-  let add env vars id body loc =
-    if MString.mem id env.to_ty then error (ClashType id) loc;
-    let ty_vars = fresh_vars env vars loc in
+  let add ?(recursive=false) env vars id body loc =
+    if MString.mem id env.to_ty && not recursive then error (ClashType id) loc;
+    let ty_vars = fresh_vars ~recursive env vars loc in
     match body with
     | Abstract ->
       let ty = Ty.text ty_vars id in
       ty, { env with to_ty = MString.add id ty env.to_ty }
     | Enum lc ->
       let ty = Ty.tsum id lc in
       ty, { env with to_ty = MString.add id ty env.to_ty }
-    | Record lbs ->
+    | Record (record_constr, lbs) ->
       let lbs =
         List.map (fun (x, pp) -> x, ty_of_pp loc env None pp) lbs in
-      let ty = Ty.trecord ty_vars id lbs in
+      let ty = Ty.trecord ~record_constr ty_vars id lbs in
       ty, { to_ty = MString.add id ty env.to_ty;
             from_labels =
               List.fold_left
                 (fun fl (l,_) -> MString.add l id fl) env.from_labels lbs }
+    | Algebraic l
+        [@ocaml.ppwarning "TODO: check non duplication ? other things"]
+
+      ->
+      let l =
+        List.map (fun (constr, l) ->
+            constr
+              [@ocaml.ppwarning "TODO: register selectors in env"],
```

Comment:
Is that TODO blocking ? (i.e is it needed before the PR is merged ?)
Are selectors the same things as the projection operators (aka field access) ? Or are they used for the `isConstr` cases in the AST ?